kame/kame/cksum6/cksum6.1
.\"	$KAME: cksum6.1,v 1.2 2000/11/09 03:04:06 itojun Exp $
.\"
.\" Copyright (C) 1995, 1996, 1997, and 1998 WIDE Project.
.\" All rights reserved.
.\" 
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the project nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE PROJECT AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE PROJECT OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.Dd May 17, 1998
.Dt CKSUM6 1
.Os KAME
.Sh NAME
.Nm cksum6
.Nd compute IPv6 pseudo-header checksum
.Sh SYNOPSIS
.Nm
.Op Fl ad
.Sh DESCRIPTION
.Nm Cksum6
will compute internet standard checksum for the datastream
given from the standard input, as hexadecimal dump form
.Po
such as output of
.Xr od 1
.Pc .
.Bl -tag -width Ds
.It Fl a
By default,
.Nm
will treat the input packet as TCP6/UDP6 packet with IPv6 header,
and IPv6 header part (40 bytes) will be skipped.
With
.Fl a
option,
.Nm
will compute checksum for the whole datastream.
.It Fl d
Debug mode.
.El
.Sh RETURN VALUES
.Nm Cksum6
exits with 0 on success, and non-zero with errors.
.Sh SEE ALSO
.Xr od 1
.Sh HISTORY
The
.Nm
command first appeared in WIDE Hydrangea IPv6 protocol stack kit.
.\" .Sh BUGS
.\" (to be written)


CHANGELOG
$KAME: CHANGELOG,v 1.2458 2003/08/18 05:39:09 itojun Exp $

<200308>
Sun Aug 17 JST 2003  itojun@iijlab.net
	* sys/netinet{,6}/in{,6}_proto.c: enforce inbound IPsec policy on
	  protocols that falls into raw socket as a wildcard case.

2003-08-15  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/kame/pim6sd/{cfparse.y, mrt.[ch], pim6_proto.c, rp.c}
	Now static-RP configuration can work together with bootstrap-
	message-based RP configuration

2003-08-10  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/{freebsd5, kame}
	sync with FreeBSD-5.1 RELEASE (behaviour is not confirmed yet)

2003-08-07  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* kame/kame/had/mpa.c,haadisc.c
	A home address should be passed in the source address of a mobile
	solicit pakcet. This problem was reported and donated the code by
	Kenichi Yajima <yajima@netlab.nec.co.jp>
	
2003-08-07  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/cftoken.l: use a stronger lexical check for
	interface names.
	From: Hajimu UMEMOTO <ume@mahoroba.org>
	* kame/kame/dhcp6/config.c (configure_interface): check interfaces
	specified in the configuration file really exist.

2003-08-05  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* removed vrrp6 support because it might be a patent infringement.
	  http://www.ietf.org/ietf/IPR/cisco-ipr-draft-ietf-vrrp-ipv6-spec.txt

2003-08-01  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: implemented various clarifications and some new
	  features:
	- supported DNS name and NTP servers options
	- dhcp6s now uses the server/relay port to send a relay-rely
	  message
	- dhcp6s now ignores a rebind message when it cannot find a
	  binding, according to prefix-delegation-04
	- dhcp6relay now accepts relay-reply messages on the socket
	  listening on the server/relay port
	* The DHCPv6 implementation is now built by default under
	  *bsd/usr.sbin/.

<200307>
Thu Jul 31 19:19:48 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c,mip6_cncore.c
	sending/receiving a binding refresh request message is supported.

Tue Jul 29 17:09:15 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.[hc]
	select a CoA using the source address selection like algorithm.
	suggested by Francis.Dupont@enst-bretagne.fr.

2003-07-28  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* sys/netinet6/{ip6_output.c, ip6_var.h, mip6_cncore.c,
	mip6_hacore.[ch], nd6.c}:
	CN or HA dosen't add a type2 routing header to a Binding
	Acknowledgement packet when a deregistration packet was issued
	on home link and it was failed by some reasons.
	reported by Yukiyo.Akisada@jp.yokogawa.com and v6pc cert.
	team.

Mon Jul 28 20:03:46 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_cncore.c,mip6_mncore.[hc],in6.c,nd6_rtr.c
	  kame/sys/net/if_hif.[hc]
	assign a CoA separately for each home interface, since a home
	interface is logically independent from each other.

Mon Jul 28 14:38:11 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_hacore.c,mip6_cncore.c:
	- fixed a bug not sending a binding ack when DAD for a link-local
	  address is failed.  (en-bugged during re-structuring BC logic.)
	- fixed a bug using RTHDR2 with Mobility Headers other than
	  a binding ack.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

Fri Jul 25 19:12:15 JST 2003  itojun@iijlab.net
	* sys/netinet/ah_core.c: hmac-ripemd160 support

Fri Jul 25 18:07:43 JST 2003 sakane@kame.net
	* kame/sys/netkey/key.c
	fixed that the kernel crashed when key_spdacquire() was called
	because key_spdacquire() had been implemented imcopletely.

Thu Jul 24 16:11:06 JST 2003	keiichi@iij.ad.jp
	* mip6 related files.
	simplify the prefix and advertising router list management
	mechanism on a mobile node.

2003-07-20  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/config.c (configure_duid): fixed a bug of
	allocating short memory.  Upgrading is required if you configure
	dhcp6s with the duid statement.

Sat Jul 19 18:12:01 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_aesctr.c: support draft-ietf-ipsec-ciph-aes-ctr-03
	  as there's no official DOI assignment yet, we use private DOI number.
	* sys/netinet6/ah_aesxcbcmac.c: support
	  draft-ietf-ipsec-ciph-aes-xcbc-mac-03.
	  as there's no official DOI assignment yet, we use private DOI number.

Tue Jul 15 20:02:05 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_rijndael.c: simplify and update rijndael code.
	  markus@openbsd

2003-07-14  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: revised the relay agent implementation
	based on dhcpv6-28.
	- dhcp6relay now relays packets between clients and servers using
	  relay forward/reply messages.
	- dhcp6s now accepts relay forward messages and replies with relay
	  reply messages.

Sat Jul 12 15:05:23 JST 2003	suz@crl.hitachi.co.jp
	* openbsd/sys/netinet/igmp_var.h
	makes IGMPV3-kernel compilable on openbsd.

Thu Jul 10 21:35:04 JST 2003	keiichi@iij.ad.jp
	* kame/kame/had/halist.c
	fixed a bug that a DHAAD reply message only includes one global
	address per home agent.  the message must include all global
	addresses of all home agents.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

2003/07/10 21:07:50 JST kjc@csl.sony.co.jp
	the first step for transition to pf/altq.
	marge pf/altq into the existing altq-3.
	allow both pf/altq and altq-3 to coexit.
	currently, ALTQ3_COMPAT and ALTQ3_CLFIER_COMPAT are defined
	in sys/altq/altq.h to support altq-3.
	altq-3 will remain in kame for research experiments but will
	not be merged into bsd releases.

2003-07-08  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* kame/sys/netinet6/{in6_ifattach.c,ip6_mroute.c,ip6_mroute.h}
	protect against interface removal.

Tue Jul  8 18:51:06 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c
	insert alternate careof address sub-option when creating a
	binding update for the home registration.  This sub-option is
	MUST when we protect the message with ESP and strictly speaking,

kame/kame/cksum6/cksum6.c
/*	$KAME: cksum6.c,v 1.3 2000/11/24 11:36:15 itojun Exp $	*/

/*
 * Copyright (c) 1989, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Mike Muuss.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/types.h>
#include <stdio.h>
#include <ctype.h>
#include <stdlib.h>
#include <unistd.h>

#define MAXPACKET	8192

u_char packet[MAXPACKET];
int tcp, udp, icmp;
int all;
int debug;

main(argc, argv)
	char *argv[];
{
	int i;
	u_int a;
	int len = 0;
	int off;
	u_long cksum, oldsum;
	unsigned char buf[BUFSIZ], *p, *np;
	int ch;

	while ((ch = getopt(argc, argv, "ad")) != EOF) {
		switch (ch) {
		case 'a':
			all++;
			break;
		case 'd':
			debug++;
			break;
		}
	}
	argc -= optind;
	argv += optind;
	if (argc) {
		fprintf(stderr, "too many arguments\n");
		exit(1);
	}

	i = 0;
#if 0
	while (scanf("%2x", &a) > 0)
		packet[i++] = a;
#else
	while (fgets((char *)buf, sizeof(buf), stdin)) {
		p = buf;
		for (;;) {
			while (*p && !isxdigit(*p))
				p++;
			if (*p == '\0')
				break;
			if (sscanf((char *)p, "%2x", &a) != 1) {
				printf("error in sscanf: \"%s\"\n", p);
				exit(1);
			}
			p += 2;
			packet[i++] = a;
		}
	}
#endif

	if (all) {
		off = 0;
		len = i;
	} else {
		off = 40;
		len = (packet[4] << 8) | packet[5];
	}
	switch (packet[6]) {
	case 6:
		tcp++;
		break;
	case 17:
		udp++;
		break;
	case 58:
		icmp++;
		break;
	default:
		printf("unknown proto %x\n", packet[6]);
		break;
	}
	if (debug)
		printf("off=%d, len=%d, input=%d, proto=%d\n", off, len, i,
		    packet[6]);
	if (all)
		cksum = in_cksum(packet, len);
	else {
		u_char ipovly[40];
		bcopy(&packet[8], &ipovly[0], 32);
		ipovly[32] = 0;
		ipovly[33] = 0;
		ipovly[34] = packet[4];
		ipovly[35] = packet[5];
		ipovly[36] = 0;
		ipovly[37] = 0;
		ipovly[38] = 0;
		ipovly[39] = packet[6];
		bcopy(ipovly, (packet + off) - sizeof(ipovly), sizeof(ipovly));
		if (tcp) {
			oldsum = (packet[off + 16] << 8) + packet[off + 17];
			packet[off + 16] = packet[off + 17] = 0; /* clear check sum */
		} else if (udp) {
			oldsum = (packet[off + 6] << 8) + packet[off + 7];
			packet[off + 6] = packet[off + 7] = 0; /* clear check sum */
		} else if (icmp) {
			oldsum = (packet[off + 2] << 8) + packet[off + 3];
			packet[off + 2] = packet[off + 3] = 0; /* clear check sum */
		} else {
			printf("XXX\n");
			abort();
		}
		if (i < len)
			printf("SHORT PACKET!!\n");
		cksum = htons(in_cksum((packet + off) - sizeof(ipovly),
				       len + sizeof(ipovly)));
	}
	if (all)
		printf("checksum = %04x (~%04x)\n", cksum, (u_short)~cksum);
	else if (oldsum == cksum)
		printf("checksum = %04x\n", cksum);
	else
		printf("BAD checksum: %04x != %04x\n", oldsum, cksum);
	exit(0);
}


/*
 * in_cksum --
 *	Checksum routine for Internet Protocol family headers (C Version)
 */
in_cksum(addr, len)
	u_short *addr;
	int len;
{
	register int nleft = len;
	register u_short *w = addr;
	register int sum = 0;
	u_short answer = 0;

	if (debug) {
		u_char *p = (u_char *)addr;
		int i;
		printf("checksum packets: \n\t");
		for (i = 0; i < len; i++) {
			printf("%02x", p[i]);
			if (i % 16 == 15)
				printf("\n\t");
			else if (i % 2 == 1)
				printf(" ");
		}
		printf("\n");
	}
	/*
	 * Our algorithm is simple, using a 32 bit accumulator (sum), we add
	 * sequential 16 bit words to it, and at the end, fold back all the
	 * carry bits from the top 16 bits into the lower 16 bits.
	 */
	while (nleft > 1)  {
		sum += *w++;
		nleft -= 2;
	}

	/* mop up an odd byte, if necessary */
	if (nleft == 1) {
		*(u_char *)(&answer) = *(u_char *)w ;
		sum += answer;
	}

	/* add back carry outs from top 16 bits to low 16 bits */
	sum = (sum >> 16) + (sum & 0xffff);	/* add hi 16 to low 16 */
	sum += (sum >> 16);			/* add carry */
	answer = ~sum;				/* truncate to 16 bits */
	return(answer);
}
